theano/tensor/signal/pool.py

```diff
@@ -310,23 +366,25 @@ def perform(self, node, inp, out):
         elif self.mode != 'max':
             func = numpy.average
 
-        for n in xrange(x.shape[0]):
-            for k in xrange(x.shape[1]):
-                for r in xrange(pr):
-                    row_st = r * st0
-                    row_end = builtins.min(row_st + ds0, img_rows)
-                    if not inc_pad:
-                        row_st = builtins.max(row_st, self.padding[0])
-                        row_end = builtins.min(row_end, x.shape[-2] + pad_h)
-                    for c in xrange(pc):
-                        col_st = c * st1
-                        col_end = builtins.min(col_st + ds1, img_cols)
-                        if not inc_pad:
-                            col_st = builtins.max(col_st, self.padding[1])
-                            col_end = builtins.min(col_end,
-                                                   x.shape[-1] + pad_w)
-                        zz[n, k, r, c] = func(y[
-                            n, k, row_st:row_end, col_st:col_end])
+        # precompute the region boundaries for each dimension
+        region_slices = [[] for i in xrange(nd)]
+        for i in xrange(nd):
+            for j in xrange(pool_out_shp[i]):
+                start = j * st[i]
+                end = builtins.min(start + ds[i], img_shp[i])
+                if not inc_pad:
+                    start = builtins.max(start, pad[i])
+                    end = builtins.min(end, img_shp[i] - pad[i])
+                region_slices[i].append(slice(start, end))
+
+        # iterate over non-pooling dimensions
```

Comment:
You seem to implement that now, The Pool op can pool over an inputs that is more then 4 or 5 dims and do that in the op instead of doing reshape in the graph. Why did you bother with that? The reshape op is free when the colapsed dimensions are contiguous in memory. Which is most of the time.

I'm fine to keep that. I'm more interrested to know your motivation for that.